src/Storefront/Theme/ThemeService.php

```diff
@@ -284,38 +285,86 @@ public function getThemeConfiguration(string $themeId, bool $translate, Context
      */
     public function getThemeConfigurationStructuredFields(string $themeId, bool $translate, Context $context): array
     {
-        $mergedConfig = $this->getThemeConfiguration($themeId, $translate, $context)['fields'];
+        $themeConfig = $this->getThemeConfiguration($themeId, $translate, $context);
+        $themeTechnicalName = $themeConfig['themeTechnicalName'];
+        $mergedFieldConfig = $themeConfig['fields'];
 
         $translations = [];
         if ($translate) {
             $translations = $this->getTranslations($themeId, $context);
-            $mergedConfig = $this->translateLabels($mergedConfig, $translations);
+            $mergedFieldConfig = $this->translateLabels($mergedFieldConfig, $translations);
         }
 
         $outputStructure = [];
 
-        foreach ($mergedConfig as $fieldName => $fieldConfig) {
+        foreach ($mergedFieldConfig as $fieldName => $fieldConfig) {
             $tab = $this->getTab($fieldConfig);
             $tabLabel = $this->getTabLabel($tab, $translations);
+            $tabSnippetKey = $this->buildSnippetKey($themeTechnicalName, false, $tab);
+
             $block = $this->getBlock($fieldConfig);
             $blockLabel = $this->getBlockLabel($block, $translations);
+            $blockSnippetKey = $this->buildSnippetKey($themeTechnicalName, false, $tab, $block);
+
             $section = $this->getSection($fieldConfig);
             $sectionLabel = $this->getSectionLabel($section, $translations);
+            $sectionSnippetKey = $this->buildSnippetKey($themeTechnicalName, false, $tab, $block, $section);
 
             // set default tab
             $outputStructure['tabs']['default']['label'] = '';
 
             // set labels
             $outputStructure['tabs'][$tab]['label'] = $tabLabel;
+            $outputStructure['tabs'][$tab]['labelSnippetKey'] = $tabSnippetKey;
             $outputStructure['tabs'][$tab]['blocks'][$block]['label'] = $blockLabel;
+            $outputStructure['tabs'][$tab]['blocks'][$block]['labelSnippetKey'] = $blockSnippetKey;
             $outputStructure['tabs'][$tab]['blocks'][$block]['sections'][$section]['label'] = $sectionLabel;
+            $outputStructure['tabs'][$tab]['blocks'][$block]['sections'][$section]['labelSnippetKey'] = $sectionSnippetKey;
+
+            $custom = $fieldConfig['custom'] ?? null;
+
+            if ($custom && \is_array($custom['options'])) {
```

Comment:
Could you please split the logic of this function into smaller private methods? Public functions should be concise and easy to read. Using well-named private methods instead of large, nested code blocks would significantly improve readability.

Reply:
Done